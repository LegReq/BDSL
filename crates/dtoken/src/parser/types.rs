use std::collections::HashMap;

use tinyjson::JsonValue;

use super::{group::Group, token::Token};

#[derive(Debug, Clone, PartialEq)]
pub struct DesignTokens {
    pub items: HashMap<String, TokenOrGroup>,
}

impl DesignTokens {
    pub fn from_map(map: &HashMap<String, JsonValue>) -> Option<Self> {
        let items = map
            .iter()
            .filter_map(|(k, v)| Some((k.to_owned(), TokenOrGroup::from_map(v.get()?, None)?)))
            .collect();

        Some(Self { items })
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenOrGroup {
    Token(Token),
    Group(Group),
}
impl TokenOrGroup {
    pub fn from_map(
        map: &HashMap<String, JsonValue>,
        default_type: Option<String>,
    ) -> Option<Self> {
        if map.contains_key("$value") {
            Token::from_map(map, default_type).map(TokenOrGroup::Token)
        } else {
            Group::from_value(map, default_type).map(TokenOrGroup::Group)
        }
    }

    pub fn description(&self) -> Option<&str> {
        match self {
            TokenOrGroup::Token(v) => v.description.as_deref(),
            TokenOrGroup::Group(v) => v.description.as_deref(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::{
        parser::token::Value,
        types::{color::Color, dimension::Dimension},
    };

    use super::*;
    use tinyjson::JsonValue::{Object, String};

    #[test]
    fn test_design_tokens_from_map() {
        let test_cases = vec![
            (
                HashMap::from([
                    (
                        "color".to_string(),
                        Object(HashMap::from([
                            ("$type".to_string(), String("color".to_owned())),
                            ("$value".to_string(), String("#FF5733".to_owned())),
                            ("$description".to_string(), String("Red color".to_owned())),
                        ])),
                    ),
                    (
                        "dimension".to_string(),
                        Object(HashMap::from([
                            ("$type".to_string(), String("dimension".to_owned())),
                            ("$value".to_string(), String("16px".to_owned())),
                        ])),
                    ),
                ]),
                Some(DesignTokens {
                    items: vec![
                        (
                            "color".to_owned(),
                            TokenOrGroup::Token(Token {
                                value: Value::Color(Color {
                                    r: 255,
                                    g: 87,
                                    b: 51,
                                    a: 255,
                                }),
                                description: Some("Red color".to_owned()),
                            }),
                        ),
                        (
                            "dimension".to_owned(),
                            TokenOrGroup::Token(Token {
                                value: Value::Dimension(Dimension::Pixels(16.0)),
                                description: None,
                            }),
                        ),
                    ]
                    .into_iter()
                    .collect(),
                }),
            ),
            (
                HashMap::from([(
                    "group".to_string(),
                    Object(HashMap::from([
                        (
                            "subgroup".to_string(),
                            Object(HashMap::from([
                                ("$type".to_string(), String("color".to_owned())),
                                ("$value".to_string(), String("#00FF00".to_owned())),
                            ])),
                        ),
                        ("$type".to_string(), String("dimension".to_owned())),
                    ])),
                )]),
                Some(DesignTokens {
                    items: vec![(
                        "group".to_owned(),
                        TokenOrGroup::Group(Group {
                            items: vec![(
                                "subgroup".to_owned(),
                                TokenOrGroup::Token(Token {
                                    value: Value::Color(Color {
                                        r: 0,
                                        g: 255,
                                        b: 0,
                                        a: 255,
                                    }),
                                    description: None,
                                }),
                            )]
                            .into_iter()
                            .collect(),
                            description: None,
                            default_type: Some("dimension".to_owned()),
                            extensions: HashMap::new(),
                        }),
                    )]
                    .into_iter()
                    .collect(),
                }),
            ),
        ];

        for (input, expected) in test_cases {
            let result = DesignTokens::from_map(&input);
            assert_eq!(result, expected);
        }
    }

    #[test]
    fn test_token_or_group_from_map() {
        // Test cases for Token variant
        let token_test_cases = vec![
            (
                HashMap::from([
                    ("$type".to_string(), String("color".to_owned())),
                    ("$value".to_string(), String("#FF5733".to_owned())),
                ]),
                None,
                Some(TokenOrGroup::Token(Token {
                    value: Value::Color(Color {
                        r: 255,
                        g: 87,
                        b: 51,
                        a: 255,
                    }),
                    description: None,
                })),
            ),
            (
                HashMap::from([
                    ("$type".to_string(), String("dimension".to_owned())),
                    ("$value".to_string(), String("16px".to_owned())),
                ]),
                None,
                Some(TokenOrGroup::Token(Token {
                    value: Value::Dimension(Dimension::Pixels(16.0)),
                    description: None,
                })),
            ),
        ];

        for (input, default_type, expected) in token_test_cases {
            let result = TokenOrGroup::from_map(&input, default_type);
            similar_asserts::assert_eq!(result, expected);
        }

        // Test cases for Group variant
        let group_test_cases = vec![(
            HashMap::from([
                ("group".to_string(), Object(HashMap::new())),
                ("$type".to_string(), String("dimension".to_owned())),
            ]),
            None,
            Some(TokenOrGroup::Group(Group {
                items: HashMap::from([(
                    "group".to_string(),
                    TokenOrGroup::Group(Group {
                        items: HashMap::new(),
                        description: None,
                        default_type: Some("dimension".to_owned()),
                        extensions: HashMap::new(),
                    }),
                )]),
                description: None,
                default_type: Some("dimension".to_owned()),
                extensions: HashMap::new(),
            })),
        )];

        for (input, default_type, expected) in group_test_cases {
            let result = TokenOrGroup::from_map(&input, default_type);
            similar_asserts::assert_eq!(result, expected);
        }
    }
}
