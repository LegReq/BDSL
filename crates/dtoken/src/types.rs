pub mod alias;
pub mod border;
pub mod color;
pub mod cubic_bezier;
pub mod dimension;
pub mod duration;
pub mod font_family;
pub mod font_weight;
pub mod gradient;
pub mod number;
pub mod shadow;
pub mod stroke_style;
pub mod transition;
pub mod typography;
