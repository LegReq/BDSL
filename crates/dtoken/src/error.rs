//! Errors returned when generating code.

use std::{
    fmt::{self, Display},
    num::{ParseFloatError, ParseIntError},
};

use tinyjson::JsonParseError;

/// Error type returned when the code generation failed for some reason.
#[derive(Debug)]
pub enum BuildError {
    Parse(Error),
    Fmt(std::io::Error),
    Read(std::io::Error),
    Write(std::io::Error),
    Var(std::env::VarError),
    JsonParse(tinyjson::JsonParseError),
    #[cfg(feature = "toml")]
    TomlParse(toml_span::Error),
    #[cfg(feature = "ason")]
    AsonParse(ason::AsonError),
    #[cfg(feature = "jsonc")]
    JsoncParse(jsonc_parser::errors::ParseError),
}

impl std::error::Error for BuildError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            BuildError::Parse(v) => Some(v),
            BuildError::Var(v) => Some(v),
            BuildError::JsonParse(v) => Some(v),
            #[cfg(feature = "toml")]
            BuildError::TomlParse(v) => Some(v),
            #[cfg(feature = "ason")]
            BuildError::AsonParse(v) => Some(v),
            #[cfg(feature = "jsonc")]
            BuildError::JsoncParse(v) => Some(v),
            BuildError::Fmt(v) | BuildError::Read(v) | BuildError::Write(v) => Some(v),
        }
    }
}

impl Display for BuildError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            BuildError::Parse(error) => write!(f, "failed to parse file: {error}"),
            BuildError::Fmt(error) => write!(f, "failed to run rustfmt: {error}"),
            BuildError::Read(error) => write!(f, "failed to read file: {error}"),
            BuildError::Write(error) => write!(f, "failed to write file: {error}"),
            BuildError::Var(error) => write!(f, "failed to read environment variable: {error}"),
            BuildError::JsonParse(error) => write!(f, "failed to parse json file: {error}"),
            #[cfg(feature = "toml")]
            BuildError::TomlParse(error) => write!(f, "failed to parse toml file: {error}"),
            #[cfg(feature = "ason")]
            BuildError::AsonParse(error) => write!(f, "failed to parse ason file: {error}"),
            #[cfg(feature = "jsonc")]
            BuildError::JsoncParse(error) => write!(f, "failed to parse jsonc file: {error}"),
        }
    }
}

impl From<Error> for BuildError {
    fn from(value: Error) -> Self {
        Self::Parse(value)
    }
}

impl From<JsonParseError> for BuildError {
    fn from(source: JsonParseError) -> Self {
        Self::JsonParse(source)
    }
}

#[cfg(feature = "toml")]
impl From<toml_span::Error> for BuildError {
    fn from(source: toml_span::Error) -> Self {
        Self::TomlParse(source)
    }
}

#[cfg(feature = "ason")]
impl From<ason::AsonError> for BuildError {
    fn from(source: ason::AsonError) -> Self {
        Self::AsonParse(source)
    }
}

#[cfg(feature = "jsonc")]
impl From<jsonc_parser::errors::ParseError> for BuildError {
    fn from(source: jsonc_parser::errors::ParseError) -> Self {
        Self::JsoncParse(source)
    }
}

impl From<std::env::VarError> for BuildError {
    fn from(error: std::env::VarError) -> Self {
        Self::Var(error)
    }
}

/// Error type returned when a parsing error occurs.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Error {
    Property(&'static str, Box<Error>),
    Kind(String, Box<Error>),

    MustExist,

    ExpectedString,
    ExpectedNumber,
    ExpectedArray,
    ExpectedObject,
    UnexpectedType,

    CollectionEmpty,
    CollectionLength(usize),
    ExpectedItemString,
    ExpectedItemNumber,
    ExpectedItemObject,

    NumberMustBePositive,
    NumberWithin(isize, isize),

    InvalidNumber(String),
    InvalidUnit(&'static [&'static str]),
    InvalidFormat(&'static str),
    MissingToken(char),
}

impl Error {
    #[must_use]
    pub fn prop(property: &'static str, error: Self) -> Self {
        Self::Property(property, Box::new(error))
    }

    #[must_use]
    pub fn kind(kind: String, err: Self) -> Self {
        Self::Kind(kind, Box::new(err))
    }
}

impl From<ParseFloatError> for Error {
    fn from(err: ParseFloatError) -> Self {
        Self::InvalidNumber(err.to_string())
    }
}

impl From<ParseIntError> for Error {
    fn from(err: ParseIntError) -> Self {
        Self::InvalidNumber(err.to_string())
    }
}

impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Error::Property(_, source) => Some(source.as_ref()),
            _ => None,
        }
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Property(prop, err) => write!(f, "property '{prop}' error: {err}"),
            Self::Kind(kind, err) => write!(f, "value error for $type '{kind}': {err}"),
            Self::MustExist => write!(f, "must exist"),
            Self::ExpectedString
            | Self::ExpectedNumber
            | Self::ExpectedArray
            | Self::ExpectedObject => {
                let ty = match self {
                    Self::ExpectedString => "string",
                    Self::ExpectedNumber => "number",
                    Self::ExpectedArray => "array",
                    _ => unreachable!(),
                };

                write!(f, "must be of type {ty}")
            }
            Self::ExpectedItemString | Self::ExpectedItemNumber | Self::ExpectedItemObject => {
                let ty = match self {
                    Self::ExpectedItemString => "string",
                    Self::ExpectedItemNumber => "number",
                    Self::ExpectedItemObject => "object",
                    _ => unreachable!(),
                };

                write!(f, "collection items must be of type {ty}")
            }
            Self::MissingToken(char) => write!(f, "missing token: {char}"),
            Self::NumberWithin(start, end) => write!(f, "number must be >= {start} <= {end}"),
            Self::UnexpectedType => write!(f, "unexpected type"),
            Self::NumberMustBePositive => write!(f, "must be a positive number"),
            Self::InvalidNumber(err) => write!(f, "invalid number format: {err}"),
            Self::InvalidUnit(units) => {
                write!(f, "invalid unit, must be one of {}", units.join(", "))
            }
            Self::InvalidFormat(str) => write!(f, "invalid format: {str}"),
            Self::CollectionEmpty => write!(f, "collection must not be empty"),
            Self::CollectionLength(len) => write!(f, "collection must contain {len} elements"),
        }
    }
}
