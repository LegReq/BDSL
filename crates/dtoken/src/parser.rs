pub mod group;
pub mod token;
pub mod types;
