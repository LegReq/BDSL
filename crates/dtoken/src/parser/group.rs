use std::collections::HashMap;

use tinyjson::JsonValue;

use super::types::TokenOrGroup;

#[derive(Debug, Clone, PartialEq)]
pub struct Group {
    pub items: HashMap<String, TokenOrGroup>,
    pub description: Option<String>,
    pub default_type: Option<String>,
    pub extensions: HashMap<String, JsonValue>,
}

impl Group {
    pub fn from_value(
        map: &HashMap<String, JsonValue>,
        mut default_type: Option<String>,
    ) -> Option<Self> {
        let mut items = HashMap::new();
        let mut description = None;
        let mut extensions = HashMap::new();

        if let Some(kind) = map.get("$type").and_then(|v| v.get::<String>()) {
            if !Self::is_valid_type(&kind) {
                return None; // Invalid type value
            }

            default_type = Some(kind.clone());
        }

        for (key, val) in map {
            match (key.as_str(), val) {
                ("$description", JsonValue::String(desc)) => {
                    description = Some(desc.clone());
                }
                ("$extensions", JsonValue::Object(exts)) => {
                    extensions.extend(exts.clone());
                }
                ("$type", JsonValue::String(_)) => { /* already covered */ }
                (_, JsonValue::Object(map)) => {
                    let item = TokenOrGroup::from_map(map, default_type.clone())?;
                    items.insert(key.clone(), item);
                }
                _ => return None,
            }
        }

        Some(Group {
            items,
            description,
            default_type,
            extensions,
        })
    }

    fn is_valid_type(type_str: &str) -> bool {
        let valid_types = vec![
            "border",
            "color",
            "cubicBezier",
            "dimension",
            "duration",
            "fontFamily",
            "fontWeight",
            "gradient",
            "number",
            "shadow",
            "strokeStyle",
            "transition",
            "typography",
        ];
        valid_types.contains(&type_str)
    }
}

#[cfg(test)]
mod tests {
    use crate::{
        parser::token::{Token, Value},
        types::{color::Color, dimension::Dimension},
    };

    use super::*;
    use tinyjson::JsonValue::{Number, Object, String};

    #[test]
    fn test_group_from_value() {
        let test_cases = vec![(
            HashMap::from([
                (
                    "group1".to_string(),
                    Object(HashMap::from([
                        ("$type".to_string(), String("color".to_owned())),
                        ("$value".to_string(), String("#FF5733".to_owned())),
                    ])),
                ),
                (
                    "group2".to_string(),
                    Object(HashMap::from([
                        ("$type".to_string(), String("dimension".to_owned())),
                        ("$value".to_string(), String("16px".to_owned())),
                    ])),
                ),
                (
                    "nested_group".to_string(),
                    Object(HashMap::from([
                        (
                            "nested_group1".to_string(),
                            Object(HashMap::from([
                                ("$type".to_string(), String("color".to_owned())),
                                ("$value".to_string(), String("#00FF00".to_owned())),
                            ])),
                        ),
                        (
                            "nested_group2".to_string(),
                            Object(HashMap::from([(
                                "$value".to_string(),
                                String("32px".to_owned()),
                            )])),
                        ),
                        ("$type".to_string(), String("dimension".to_owned())),
                    ])),
                ),
                (
                    "$description".to_string(),
                    String("A group of tokens".to_owned()),
                ),
                (
                    "$extensions".to_string(),
                    Object(HashMap::from([("key1".to_string(), Number(42.0))])),
                ),
            ]),
            Some(Group {
                items: vec![
                    (
                        "group1".to_owned(),
                        TokenOrGroup::Token(Token {
                            value: Value::Color(Color {
                                r: 255,
                                g: 87,
                                b: 51,
                                a: 255,
                            }),
                            description: None,
                        }),
                    ),
                    (
                        "group2".to_owned(),
                        TokenOrGroup::Token(Token {
                            value: Value::Dimension(Dimension::Pixels(16.0)),
                            description: None,
                        }),
                    ),
                    (
                        "nested_group".to_owned(),
                        TokenOrGroup::Group(Group {
                            items: vec![
                                (
                                    "nested_group1".to_owned(),
                                    TokenOrGroup::Token(Token {
                                        value: Value::Color(Color {
                                            r: 0,
                                            g: 255,
                                            b: 0,
                                            a: 255,
                                        }),
                                        description: None,
                                    }),
                                ),
                                (
                                    "nested_group2".to_owned(),
                                    TokenOrGroup::Token(Token {
                                        value: Value::Dimension(Dimension::Pixels(32.0)),
                                        description: None,
                                    }),
                                ),
                            ]
                            .into_iter()
                            .collect(),
                            description: None,
                            default_type: Some("dimension".to_owned()),
                            extensions: HashMap::new(),
                        }),
                    ),
                ]
                .into_iter()
                .collect(),
                description: Some("A group of tokens".to_owned()),
                default_type: None,
                extensions: HashMap::from([("key1".to_owned(), Number(42.0))]),
            }),
        )];

        for (input, expected) in test_cases {
            let result = Group::from_value(&input, None);
            similar_asserts::assert_eq!(result, expected);
        }
    }
}
