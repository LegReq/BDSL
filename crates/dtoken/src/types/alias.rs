//! Instead of having explicit values, tokens can reference the value of another token. To put it
//! another way, a token can be an alias for another token. This spec considers the terms "alias"
//! and "reference" to be synonyms and uses them interchangeably.
//!
//! Aliases are useful for:
//!
//! - Expressing design choices
//! - Eliminating repetition of values in token files (DRYing up the code)
//!
//! For a design token to reference another, its value MUST be a string containing the
//! period-separated (.) path to the token it's referencing enclosed in curly brackets.
//!
//! For example:
//! Example 15
//!
//! ```json,ignore
//! {
//!   "group name": {
//!     "token name": {
//!       "$value": 1234,
//!       "$type": "number"
//!     }
//!   },
//!   "alias name": {
//!     "$value": "{group name.token name}"
//!   }
//! }
//! ```
//!
//! When a tool needs the actual value of a token it MUST resolve the reference - i.e. lookup the
//! token being referenced and fetch its value. In the above example, the "alias name" token's
//! value would resolve to 1234 because it references the token whose path is {group name.token
//! name} which has the value 1234.
//!
//! Tools SHOULD preserve references and therefore only resolve them whenever the actual value
//! needs to be retrieved. For instance, in a design tool, changes to the value of a token being
//! referenced by aliases SHOULD be reflected wherever those aliases are being used.
//!
//! Aliases MAY reference other aliases. In this case, tools MUST follow each reference until they
//! find a token with an explicit value. Circular references are not allowed. If a design token
//! file contains circular references, then the value of all tokens in that chain is unknown and an
//! appropriate error or warning message SHOULD be displayed to the user.

use std::str::FromStr;

use crate::error::Error;

#[derive(Debug, Clone, PartialEq)]
pub struct Alias {
    pub(crate) path_segments: Vec<String>,
}

impl FromStr for Alias {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if !s.starts_with('{') {
            return Err(Error::MissingToken('{'));
        }
        if !s.ends_with('}') {
            return Err(Error::MissingToken('}'));
        }
        if s.len() < 3 {
            return Err(Error::InvalidFormat("empty alias"));
        }

        let path = &s[1..s.len() - 1]; // Remove the curly braces
        let path_segments = path.split('.').map(String::from).collect();
        Ok(Self { path_segments })
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::*;

    #[test]
    fn test_alias_from_str() {
        #[rustfmt::skip]
        let test_cases = vec![
            ("{foo.bar}", Ok(Alias { path_segments: vec!["foo".to_string(), "bar".to_string()] })),
            ("{abc.xyz}", Ok(Alias { path_segments: vec!["abc".to_string(), "xyz".to_string()] })),
            ("{token}", Ok(Alias { path_segments: vec!["token".to_string()] })),
            ("not_an_alias", Err(Error::MissingToken('{'))),
            ("{}valid{}", Ok(Alias { path_segments: vec!["}valid{".to_string()] })),
            ("{foo.bar", Err(Error::MissingToken('}'))),
            ("foo.bar}", Err(Error::MissingToken('{'))),
        ];

        for (input, expected) in test_cases {
            let result = Alias::from_str(input);
            assert_eq!(result, expected);
        }
    }
}
