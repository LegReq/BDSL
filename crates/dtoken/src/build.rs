use std::{collections::HashMap, path::Path};

use crate::error::BuildError;
use crate::parser::{
    group::Group,
    token::Value,
    types::{DesignTokens, TokenOrGroup},
};
use crate::types::{
    alias::Alias, color::Color, dimension::Dimension, font_family::FontFamily, number::Number,
};
use convert_case::{Case, Casing};
use proc_macro2::{Ident, Span, TokenStream};
use quote::quote;

pub fn build(path: impl AsRef<str>) -> Result<(), BuildError> {
    use tinyjson::JsonValue;

    let content = std::fs::read_to_string(path.as_ref())?;
    let json: JsonValue = content.parse()?;
    let map = json.get().ok_or(BuildError::Parse)?;

    let tokens = DesignTokens::from_map(map).ok_or(BuildError::Parse)?;
    let code = generate(tokens);

    let output = Path::new(&std::env::var("OUT_DIR")?).join("design_tokens.rs");

    std::fs::write(&output, code.to_string())?;
    rustfmt(&output)?;

    Ok(())
}

fn generate(tokens: DesignTokens) -> TokenStream {
    Generator::new(tokens).generate()
}

struct Generator {
    root: Group,
}

impl Generator {
    fn new(tokens: DesignTokens) -> Self {
        let root = Group {
            items: tokens.items.clone(),
            description: Some("Root-level Design Tokens type".to_owned()),
            default_type: None,
            extensions: HashMap::new(),
        };

        Self { root }
    }

    fn generate(&self) -> TokenStream {
        let module = self.module_impl("DesignTokens", &self.root);
        let instance = self.group_instance("DesignTokens", &self.root, vec![]);

        quote! {
            pub fn design_tokens() -> design_tokens::DesignTokens {
                #instance
            }

            #module
        }
    }

    fn group_instance(&self, item: &str, group: &Group, mut parents: Vec<Ident>) -> TokenStream {
        let module_name = Ident::new(&item.to_case(Case::Snake), Span::call_site());
        let group_name = Ident::new(&item.to_case(Case::Pascal), Span::call_site());
        parents.push(module_name.clone());

        let mut items: Vec<_> = group.items.iter().collect();
        items.sort_by_key(|(k, _)| k.to_owned());

        let mut fields = vec![];
        let mut values = vec![];
        for (name, token_or_group) in &items {
            let (field, value) = self.field_instance(name, token_or_group, parents.clone());

            fields.push(field);
            values.push(value);
        }

        quote! {
            #(#parents::)* #group_name {
                #( #fields: #values,)*
            }
        }
    }

    fn field_instance(
        &self,
        field: &String,
        kind: &TokenOrGroup,
        parents: Vec<Ident>,
    ) -> (Ident, TokenStream) {
        let key = self.field_ident(field);
        let value = match kind {
            TokenOrGroup::Token(token) => self.token_value(&token.value),
            TokenOrGroup::Group(group) => self.group_instance(field, group, parents),
        };

        (key, value)
    }

    fn module_impl(&self, item: &str, group: &Group) -> TokenStream {
        let module = Ident::new(&item.to_case(Case::Snake), Span::call_site());
        let group = self.group_impl(item, group);

        quote! {
            #[allow(clippy::module_inception)]
            pub mod #module {
                #group
            }
        }
    }

    fn group_impl(&self, item: &str, group: &Group) -> TokenStream {
        let group_name = Ident::new(&item.to_case(Case::Pascal), Span::call_site());
        let description = group.description.clone().unwrap_or_default();

        let mut items: Vec<_> = group.items.iter().collect();
        items.sort_by_key(|(k, _)| k.to_owned());

        let mut nested = vec![];
        for (name, group_item) in &items {
            let group = self.token_or_group_impl(name, group_item);
            nested.push(group);
        }

        let mut fields = vec![];
        let mut types = vec![];
        let mut descs = vec![];
        for (name, token_or_group) in &items {
            let (field, kind) = self.struct_field(name, token_or_group);
            let desc = token_or_group.description().unwrap_or_default();

            fields.push(field);
            types.push(kind);
            descs.push(desc);
        }

        quote! {
            #[doc = #description]
            #[derive(Debug)]
            pub struct #group_name {
                #(
                #[doc = #descs]
                pub #fields: #types,
                )*
            }

            #(#nested)*
        }
    }

    fn token_or_group_impl(&self, item: &str, token_or_group: &TokenOrGroup) -> TokenStream {
        match token_or_group {
            TokenOrGroup::Token(_) => return quote! {},
            TokenOrGroup::Group(group) => self.module_impl(item, group),
        }
    }

    fn struct_field(&self, field: &String, kind: &TokenOrGroup) -> (Ident, TokenStream) {
        let key = self.field_ident(field);
        let value = match kind {
            TokenOrGroup::Token(token) => self.token_kind(&token.value),
            TokenOrGroup::Group(_) => {
                let module = Ident::new(&field.to_case(Case::Snake), Span::call_site());
                let tail = Ident::new(&field.to_case(Case::Pascal), Span::call_site());
                quote! { #module :: #tail }
            }
        };

        (key, value)
    }

    fn field_ident(&self, field: &str) -> Ident {
        let key = if field.starts_with('_') {
            format!("_{}", field.to_case(Case::Snake))
        } else {
            field.to_case(Case::Snake)
        };
        Ident::new(&key, Span::call_site())
    }

    fn alias_type(&self, alias: &Alias) -> Result<TokenStream, String> {
        let mut reference = &TokenOrGroup::Group(self.root.clone());

        for key in &alias.path_segments {
            reference = match reference {
                TokenOrGroup::Token(_) => {
                    return Err(format!(
                        "alias path segment {} points to value, but group was expected.",
                        key
                    ));
                }
                TokenOrGroup::Group(group) => match group.items.get(key) {
                    Some(token_or_group) => token_or_group,
                    None => {
                        return Err(format!(
                            "alias target missing: {{{}}}",
                            alias.path_segments.join(".")
                        ));
                    }
                },
            };
        }

        match reference {
            TokenOrGroup::Token(token) => Ok(self.token_kind(&token.value)),
            TokenOrGroup::Group(_) => Err(format!(
                "alias {{{}}} must point to a value, but instead points to a group.",
                alias.path_segments.join(".")
            )),
        }
    }

    fn token_kind(&self, value: &Value) -> TokenStream {
        let kind = match value {
            Value::Color(_) => "Color",
            Value::Dimension(_) => "Dimension",
            Value::FontFamily(_) => "FontFamily",
            Value::FontWeight(_) => "FontWeight",
            Value::Duration(_) => "Duration",
            Value::CubicBezier(_) => "CubicBezier",
            Value::Number(_) => "Number",
            Value::StrokeStyle(_) => "StrokeStyle",
            Value::Border(_) => "Border",
            Value::Transition(_) => "Transition",
            Value::Shadow(_) => "Shadow",
            Value::Gradient(_) => "Gradient",
            Value::Typography(_) => "Typography",
            Value::Alias(alias) => return self.alias_type(alias).unwrap(),
        };

        let module = Ident::new(&kind.to_case(Case::Snake), Span::call_site());
        let kind = Ident::new(kind, Span::call_site());

        quote! {
            dtoken::types::#module::#kind
        }
    }

    fn alias_value(&self, alias: &Alias) -> Result<TokenStream, String> {
        let mut reference = &TokenOrGroup::Group(self.root.clone());

        for key in &alias.path_segments {
            reference = match reference {
                TokenOrGroup::Token(_) => {
                    return Err(format!(
                        "alias path segment {} points to value, but group was expected.",
                        key
                    ));
                }
                TokenOrGroup::Group(group) => match group.items.get(key) {
                    Some(token_or_group) => token_or_group,
                    None => {
                        return Err(format!(
                            "alias target missing: {{{}}}",
                            alias.path_segments.join(".")
                        ));
                    }
                },
            };
        }

        match reference {
            TokenOrGroup::Token(token) => Ok(self.token_value(&token.value)),
            TokenOrGroup::Group(_) => Err(format!(
                "alias {{{}}} must point to a value, but instead points to a group.",
                alias.path_segments.join(".")
            )),
        }
    }

    fn token_value(&self, value: &Value) -> TokenStream {
        let kind = self.token_kind(value);
        let value = match value {
            Value::Color(Color { r, g, b, a }) => quote! { { r: #r, g: #g, b: #b, a: #a } },
            Value::Number(Number(v)) => quote! { (#v) },
            Value::Alias(alias) => return self.alias_value(alias).unwrap(),
            Value::FontFamily(FontFamily { primary, fallbacks }) => {
                quote! { {
                    primary: #primary.to_owned(),
                    fallbacks: vec![#( #fallbacks.to_owned(),)*],
                } }
            }
            Value::Dimension(v) => match v {
                Dimension::Pixels(v) => {
                    quote! { ::Pixels(#v) }
                }
                Dimension::Rems(_) => todo!(),
            },
            v => todo!("{:?}", v),
            // Value::FontWeight(v) => quote! { #v },
            // Value::Duration(v) => quote! { #v },
            // Value::CubicBezier(v) => quote! { #v },
            // Value::StrokeStyle(v) => quote! { #v },
            // Value::Border(v) => quote! { #v },
            // Value::Transition(v) => quote! { #v },
            // Value::Shadow(v) => quote! { #v },
            // Value::Gradient(v) => quote! { #v },
            // Value::Typography(v) => quote! { #v },
        };

        quote! {
            #kind #value
        }
    }
}

/// Format a file with rustfmt
#[cfg(feature = "rustfmt")]
fn rustfmt(path: &Path) -> Result<(), BuildError> {
    use std::process::Command;

    Command::new(std::env::var("RUSTFMT").unwrap_or_else(|_| "rustfmt".to_string()))
        .args(&["--emit", "files"])
        .arg(path)
        .output()
        .map_err(BuildError::Fmt)?;

    Ok(())
}

#[cfg(not(feature = "rustfmt"))]
fn rustfmt(_path: &Path) -> Result<(), BuildError> {
    Ok(())
}

#[cfg(test)]
mod tests {
    use indoc::indoc;
    use syn::File;
    use tinyjson::JsonValue;

    use super::*;

    #[test]
    fn test_examples() {
        let test_cases = vec![indoc! {r#"
                {
                  "group name": {
                    "token name": {
                      "$value": 1234,
                      "$type": "number"
                    }
                  },
                  "alias name": {
                    "$value": "{group name.token name}"
                  }
                }
            "#}];

        for case in test_cases {
            let value: JsonValue = case.parse().unwrap();
            let tokens = DesignTokens::from_map(value.get().unwrap()).unwrap();

            let tokens = generate(tokens);
            let abstract_file: File =
                syn::parse2(tokens.clone()).unwrap_or_else(|err| panic!("{err}:\n\n{}", tokens));
            let code = prettyplease::unparse(&abstract_file);

            insta::assert_snapshot!(code.to_string());
        }
    }
}
