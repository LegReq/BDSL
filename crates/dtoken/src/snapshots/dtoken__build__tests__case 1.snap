---
source: crates/dtoken/src/build.rs
expression: code.to_string()
---
#[allow(clippy::allow_attributes, clippy::too_many_lines)]
pub fn design_tokens() -> design_tokens::DesignTokens {
    design_tokens::DesignTokens {
        alias_name: dtoken::types::number::Number(1234f64),
        group_name: design_tokens::group_name::GroupName {
            token_name: dtoken::types::number::Number(1234f64),
        },
    }
}
#[allow(clippy::allow_attributes, clippy::module_inception)]
pub mod design_tokens {
    ///Root-level Design Tokens type
    #[derive(bevy_reflect::Reflect)]
    #[derive(Debug)]
    pub struct DesignTokens {
        pub alias_name: dtoken::types::number::Number,
        pub group_name: group_name::GroupName,
    }
    #[allow(clippy::allow_attributes, clippy::module_inception)]
    pub mod group_name {
        #[derive(bevy_reflect::Reflect)]
        #[derive(Debug)]
        pub struct GroupName {
            pub token_name: dtoken::types::number::Number,
        }
    }
}
